Add missing button shadows in dialogs

a1exsh opened the pull request:
Relates to #4349

a1exsh commented:
Original:
![image](https://user-images.githubusercontent.com/489601/136688492-26d1aa23-2339-4606-a23f-435feb838fab.png)

Changed:
![image](https://user-images.githubusercontent.com/489601/136688500-c4b37bd2-7fa4-43d7-90e6-22627b7b046a.png)

Branikolog commented:
Hi, @a1exsh 
Looks fantastic! :)
But I've noticed, that the original central button has its shadow (which is a part of image) merged with the frame darkening of the frame. Is it possible  to get rid of bright line (between shadow and the frame darkening) here?
![config shadow](https://user-images.githubusercontent.com/55348946/136690108-affbf808-66ba-439d-b931-1e928685adf3.png)

a1exsh commented:
@Branikolog good catch.  I think we can fix that by extending the shadow by just one more pixel (currently it is -4,4).

Branikolog commented:
@a1exsh 
Also, it feels like shadow to the left contrasts more with the background, comparing to the same place near central button. Is it just due to another background texture?

a1exsh commented:
> @a1exsh Also, it feels like shadow to the left contrasts more with the background, comparing to the same place near central button. Is it just due to another background texture?

@Branikolog that is because the `OKAY` button and the shadow beneath it are part of the original resource, but we add shadow manually for our custom button, and it doesn't match 100%.  Now that I think about it, in this case both buttons are of the same size, so what we can try is copying the original button together with the shadow instead.

Branikolog commented:
@a1exsh 
It's perfect!

a1exsh commented:
Experimental settings dialog:
![image](https://user-images.githubusercontent.com/489601/136989673-75427cff-49ec-4d35-a164-4f5828be0e55.png)

Game config dialog:
![Screenshot from 2021-10-12 18-09-08](https://user-images.githubusercontent.com/489601/136991834-a2569548-00b5-4a97-b16e-414950af6ff5.png)
![image](https://user-images.githubusercontent.com/489601/136991893-34527a4c-379c-4313-a5a9-ed9ca695d0c2.png)

Branikolog commented:
Hi, @a1exsh.
One more case found.
![image](https://user-images.githubusercontent.com/55348946/137508389-50a3f2c7-c933-43a6-9438-55bcdef2d349.png)

a1exsh commented:
Still quick (as in "hack"), but less dirty corners:
![image](https://user-images.githubusercontent.com/489601/137584004-8adb25ac-61e5-4c71-ad50-9e4a3541050a.png)

a1exsh commented:
Not sure if there are button shadows for this dialog in OG, will check later:
![image](https://user-images.githubusercontent.com/489601/139490004-fe6ee84b-cb4c-4ffd-89b9-99a6fd6464ad.png)

a1exsh commented:
There are no shadows, but I guess it might have looked a bit better with them:
![heroes2_006](https://user-images.githubusercontent.com/489601/139494498-7c05d776-cf6c-4b95-99fc-4f60bfd00d8b.png)

a1exsh commented:
Fixed gift resources buttons position and added shadows.

Before:
![image](https://user-images.githubusercontent.com/489601/139530903-b18a0345-9396-4abe-afca-3793952cc6a2.png)

After:
![Screenshot from 2021-10-30 13-15-23](https://user-images.githubusercontent.com/489601/139530780-c5995725-e348-465e-9b0e-a86c814e445e.png)

It is kinda frustrating that every dialog has its own way of calculating position and border and placing the buttons on the window...  Are there any plans to start working on a proper UI toolkit? :) @ihhub

Branikolog commented:
> Not sure if there are button shadows for this dialog in OG, will check later: ![image](https://user-images.githubusercontent.com/489601/139490004-fe6ee84b-cb4c-4ffd-89b9-99a6fd6464ad.png)

We have an opened issue for that. :)
#2261
Just not to forget to close it after fixing. ;)

a1exsh commented:
It really becomes an art at some point: every single button is unique %)

![image](https://user-images.githubusercontent.com/489601/139541872-dcace824-c92b-426a-9cad-baa349016f3b.png)

a1exsh commented on src/fheroes2/gui/ui_button.cpp:
```diff
@@ -172,6 +182,16 @@ namespace fheroes2
+            const Sprite & shadow = fheroes2::makeShadow( _getReleased(), fheroes2::Point( -4, 6 ), 3 );
+            _shadowSprite.reset( new Sprite( shadow ) );
+        }
+        return *_shadowSprite.get();
```
Not sure about this one: I've just copied the approach from `_getDisabled()`.

a1exsh commented:
While working on this I've noticed another glitch related to the Evil "show hero" / "show market" buttons used in Surrender dialog as well as when an army is willing to join the hero.

To make the pressed state we replace all back pixels (color index 36) with transform 1, but the there are pixels with the same index also in the middle of the pressed Evil buttons, not only around the border.  This results in background leaking through the drawing, which looks like this:
![image](https://user-images.githubusercontent.com/489601/139595467-e1e1d9af-73c7-4330-b610-07e849cd83bc.png)

A have a fix in the works, will either push here or open a new PR, depending on the resulting size.

a1exsh commented:
Alright, I'm now content with this change myself — please have a look when you have time. :)

Branikolog commented:
Hi, @a1exsh !
I've noticed one thing, related to shadows. While pressing "Hero info" or "Marketplace" buttons, shadow becomes more bold. While common "Yes"/"NO" buttons have same shadow in a pressed state:
![image](https://user-images.githubusercontent.com/55348946/139841396-4efff6b6-26aa-4100-820c-5139245ba9cb.png)
![image](https://user-images.githubusercontent.com/55348946/139841427-747553df-afb0-43be-b5fc-e4664a17a69e.png)
Actually, this PR becomes really huge. So this could be fixed in a separate one, if you wish. Anyway, you've done a ton of work with buttons already! Thanks!

I've also found out, that "GIFT" and "EXIT" buttons at evil interface become surrounded with gray frame once being pressed:
![image](https://user-images.githubusercontent.com/55348946/139842816-6e871f4f-f39c-41e5-af4d-4c56202f1a08.png)
The same to Surrender buttons:
![image](https://user-images.githubusercontent.com/55348946/139843139-e70e57b3-1b33-4f14-894c-99de2110118e.png)
![image](https://user-images.githubusercontent.com/55348946/139843179-fe1bf966-4d4a-4802-9fe3-92e13c9b1d01.png)
And "GIFT" window:
![image](https://user-images.githubusercontent.com/55348946/139843588-f17330d3-603e-45f0-9c11-72720b7cc849.png)
![image](https://user-images.githubusercontent.com/55348946/139843641-5d14af22-d48e-4d28-b284-f6b7034f751a.png)
A small piece of background changes while pressing OK button at setting window:
![image](https://user-images.githubusercontent.com/55348946/139844122-f6aff67f-7133-4a1f-9ae4-698d95f25325.png)
![Corner](https://user-images.githubusercontent.com/55348946/139844261-6f66aa00-c159-4df2-9f1c-8f7ab902cc51.png)
And same to evil:
![image](https://user-images.githubusercontent.com/55348946/139844786-84df98aa-6183-43bf-9cbf-f515656fd4c0.png)

a1exsh commented:
@Branikolog thanks for having a look — I have a fix for the button corners as well, just didn't push it yet to avoid bloating the PR. Will check the rest of your findings later. :)

Branikolog commented:
@a1exsh 
Thank You!

a1exsh commented:
> While pressing "Hero info" or "Marketplace" buttons, shadow becomes more bold.
 ![image](https://user-images.githubusercontent.com/55348946/139841427-747553df-afb0-43be-b5fc-e4664a17a69e.png)

Good catch — will fix it.

> I've also found out, that "GIFT" and "EXIT" buttons at evil interface become surrounded with gray frame once being pressed:
![image](https://user-images.githubusercontent.com/55348946/139842816-6e871f4f-f39c-41e5-af4d-4c56202f1a08.png)

This was the case before changes from this PR — let's fix it separately (I already know how to do that).

> The same to Surrender buttons:
![image](https://user-images.githubusercontent.com/55348946/139843139-e70e57b3-1b33-4f14-894c-99de2110118e.png)

Also was the case before, need to check how to fix it.

> And "GIFT" window:
![image](https://user-images.githubusercontent.com/55348946/139843588-f17330d3-603e-45f0-9c11-72720b7cc849.png)

Was the case, fixing seems easy.

> A small piece of background changes while pressing OK button at setting window:
![image](https://user-images.githubusercontent.com/55348946/139844122-f6aff67f-7133-4a1f-9ae4-698d95f25325.png)

Here I set a trap for myself: need to figure out why I did change the button ICN in this code :sweat_smile:

a1exsh commented:
> While pressing "Hero info" or "Marketplace" buttons, shadow becomes more bold.
> A small piece of background changes while pressing OK button at setting window:

@Branikolog fixed for both evil and good modes —- please have a look.  Let's fix other issues separately. :)

Branikolog commented:
@a1exsh 

> @Branikolog fixed for both evil and good modes —- please have a look. Let's fix other issues separately. :)

Sure!

a1exsh commented:
@Branikolog would you have time for another look on this? :)

Branikolog commented:
Hi, @a1exsh !
I'm not sure, if all the cases are related to current PR. So be free to postpone some of them to future.

I've noticed, that Creature info window has smaller shadows for buttons.
![image](https://user-images.githubusercontent.com/55348946/141099207-01d6b3c8-c1c6-47b3-91f6-05c5e769e3ae.png)
Also, unavailable "ACCEPT" button has a small black line along left and bottom border.
![image](https://user-images.githubusercontent.com/55348946/141099474-9b2b261f-00ef-40a2-bd22-d7aa63634c39.png)
Pretty sure, we haven't discussed it before, but I'm not sure it is related to the current PR.
And the same to Dig and supposedly other system buttons.
![image](https://user-images.githubusercontent.com/55348946/141101765-4d9283f3-210e-414c-b7dd-5a2e11e2c7cb.png)

a1exsh commented:
> Hi, @a1exsh ! I'm not sure, if all the cases are related to current PR. So be free to postpone some of them to future.
> 
> I've noticed, that Creature info window has smaller shadows for buttons.
![image](https://user-images.githubusercontent.com/55348946/141099207-01d6b3c8-c1c6-47b3-91f6-05c5e769e3ae.png)

Right.  This inconsistency comes from OG resources.  Not sure if we would try to make all of UI consistent one day, but at least we should be consistent within a single dialog window.

> Also, unavailable "ACCEPT" button has a small black line along left and bottom border.
![image](https://user-images.githubusercontent.com/55348946/141099474-9b2b261f-00ef-40a2-bd22-d7aa63634c39.png)
> Pretty sure, we haven't discussed it before, but I'm not sure it is related to the current PR.
> And the same to Dig and supposedly other system buttons.
![image](https://user-images.githubusercontent.com/55348946/141101765-4d9283f3-210e-414c-b7dd-5a2e11e2c7cb.png)

Nicely spotted — I can see it as well.  This is caused by our generic logic of producing disabled button sprites by applying a darkening palette.  All evil disabled buttons might suffer from it.  We might want to tackle it separately, if at all.

a1exsh commented:
@Branikolog I'd like to wrap this change up and move on to the next one: I've got a few in the pipeline, but they depend on this one being merged first. ;)

Branikolog commented:
@a1exsh 

> @Branikolog I'd like to wrap this change up and move on to the next one: I've got a few in the pipeline, but they depend on this one being merged first. ;)

Please, don't hesitate making small changes one my one, rather than merging huge part of new code.
You know, I'm always complaining on new issues being noticed. Ignore it! :D

a1exsh commented:
@ihhub please have another look when you have a minute. :)

a1exsh commented:
@ihhub maybe you could have a look on the code — the changes related to shadows are now complete, but I have more related improvements in pipeline. :)

ihhub commented on src/fheroes2/battle/battle_dialogs.cpp:
```diff
@@ -887,13 +876,17 @@ bool Battle::DialogBattleSurrender( const HeroBase & hero, u32 cost, Kingdom & k
             btnMarket.disable();
         }
         else {
+            btnMarket.drawShadow();
+            btnMarket.captureBackground();
```
Can you please explain the logic for this code? Are we drawing a shadow and then we're copying a background? If yes shouldn't it be easier to create a factory function which will accept a background image and 2 button images? The function will just return either a pair of images or a a button? We know that the background is not going to be changed.

a1exsh replied:
That's a good question: now I will have to recall everything, it has been a few weeks since I was last working on this...

ihhub replied:
I think a single function should fulfill all our needs in this greatly reducing the code complexity.

a1exsh replied:
@ihhub thanks a lot for your suggestions — I think I've addressed them all and the code is indeed much cleaner now.  I've added a special `AutoShadowButton` class to work with the shadows.  Please have another look!

ihhub reviewed (changes requested):
Hi @a1exsh , I raised few design related questions here. Would you mind to take a look at them?

a1exsh commented on src/engine/image.cpp:
```diff
@@ -1797,6 +1797,22 @@ namespace fheroes2
 
+    void ReplaceTransform( const Image & in, Image & out )
+    {
+        if ( in.empty() || out.empty() || in.singleLayer() || out.singleLayer() || in.width() != out.width() || in.height() != out.height() )
+            return;
+
+        const int32_t width = in.width();
+        const int32_t height = in.height();
+
+        const uint8_t * transformIn = in.transform();
+        const uint8_t * transformInEnd = transformIn + height * width;
+        uint8_t * transformOut = out.transform();
+        for ( ; transformIn != transformInEnd; ++transformIn, ++transformOut ) {
+            *transformOut = *transformIn;
+        }
+    }
```
Good point :sweat_smile:  I think I've had some extra logic there, but after removing that it's not really different from plain `memcpy()`.

a1exsh commented:
@Branikolog I did some code refactoring per @ihhub 's request.  To my eye everything looks good, but a slight problem remains with the pressed state of the Hero and Marketplace buttons in the Level Up and Army Joining dialogs (and that precisely because we do not draw shadows for them there, so also not capturing the background to restore it before drawing the pressed state).  I prefer to address this separately, as it is not related to adding shadows, which is the title of the PR. :)

Branikolog commented:
@a1exsh 
Sure! I believe it's much easier to work on fixes in separate PRs (even separate for each button) as well as test the changes. 
I'll check the PR once again in a few hours.

a1exsh commented:
> a slight problem remains with the pressed state of the Hero and Marketplace buttons in the Level Up and Army Joining dialogs

Hm, I just went through the changes once again and I see that this was handled by custom code before, so it can be seen as a regression.  I will still try to fix it right here, it shouldn't be too much of additional code.

a1exsh commented:
>  I will still try to fix it right here, it shouldn't be too much of additional code.

Fixed :)

a1exsh commented:
@Branikolog I've found and fixed an issue with the `OKAY` button shadow in the battle summary dialog, introduced by refactoring.
@ihhub please have another look at the code. :)

Branikolog commented:
Hi, @a1exsh !
Found an issue with "Town Portal" buttons being pressed.
![image](https://user-images.githubusercontent.com/55348946/141966367-02e740fb-095e-4061-9cf9-cc1c35fb066b.png)

a1exsh commented:
> Hi, @a1exsh ! Found an issue with "Town Portal" buttons being pressed. ![image](https://user-images.githubusercontent.com/55348946/141966367-02e740fb-095e-4061-9cf9-cc1c35fb066b.png)

If the issue you're referring to is with button corners — it is not new and I'm waiting for this PR to be merged before addressing these. I think it was discussed already in the comments here.

Branikolog commented:
Sorry, @a1exsh. I have too much tasks now. :) Missed this moment.
I was just confused by the fact, the latest commit version doesn't have these corners.

a1exsh commented:
@Branikolog 
> I was just confused by the fact, the latest commit version doesn't have these corners.

It may be, but the issue is definitely present in the latest release and it's not directly related to the shadows, so let's address it separately.

a1exsh commented:
@ihhub please find the time to have a look — I've extracted the relevant code to new subclasses of the base button class.

ihhub commented on src/engine/image.h:
```diff
@@ -267,4 +270,9 @@ namespace fheroes2
 
     void Transpose( const Image & in, Image & out );
+
+    // Generates a new image with a shadow of the shape of existing image. Shadow must have only (-x, +y) offset.
+    Sprite addShadow( const Sprite & in, const Point & shadowOffset, const uint8_t transformValue );
+
+    Sprite makeShadow( const Sprite & in, const Point & shadowOffset, const uint8_t transformValue );
```
Please put these functions to places to have alphabetical order of function names. You can ignore that these function names don't start from a capital letter.

ihhub commented on src/fheroes2/agg/agg_image.cpp:
```diff
@@ -1441,6 +1459,40 @@ namespace fheroes2
+                AddTransparency( pressed, 36 );
+                AddTransparency( pressed, 61 ); // remove the extra brown border
+
+                Sprite fixed( pressed.width(), pressed.height(), pressed.x(), pressed.y() );
```
Can we place name this variable by more understandable name? Right now `fixed` is a confusing name.

ihhub commented on src/fheroes2/agg/agg_image.cpp:
```diff
@@ -1441,6 +1459,40 @@ namespace fheroes2
+                // put back pixels that actually should be black
+                Fill( fixed, 1, 4, 31, 31, 36 );
+                Blit( pressed, fixed );
+                _icnVsSprite[id][1] = fixed;
```
Please use `std::move` instead of copying the image.

ihhub replied:
Even better: we can directly use `_icnVsSprite[id][1]` instead of `fixed`.

ihhub commented on src/fheroes2/gui/ui_button.cpp:
```diff
@@ -240,7 +240,76 @@ namespace fheroes2
     }
 
-    ButtonGroup::ButtonGroup( const Rect & area, int buttonTypes )
+    AutoBackgroundButton::AutoBackgroundButton( const Image & in, int32_t offsetX, int32_t offsetY, int icnId, uint32_t releasedIndex, uint32_t pressedIndex )
+        : ButtonSprite( offsetX, offsetY, fheroes2::AGG::GetICN( icnId, releasedIndex ), fheroes2::AGG::GetICN( icnId, pressedIndex ) )
+    {
+        _captureBackground( in );
+    }
```
This portion of the code worries me a lot. We're using `ButtonSprite` class as a base class which accepts images as constructor arguments. Here we're using ICN indexes to get images which later we copy to internal ButtonSprite members and then we're overriding them with new images within `_captureBackground` method. Also it's unknown what **in** image is for.

Name `AutoBackgroundButton` is misleading. Is it a button with automated background?

The same set of remarks is applied for `AutoShadowButton` class.

I'm suggesting to just create 2 or 4 functions, each will return ButtonSprite object. For example:
```cpp
ButtonSprite generateButtonWithShadow( int32_t offsetX, int32_t offsetY, const Sprite & released, const Sprite & pressed, const Sprite & disabled, const Image & background )
{
    // Prepare images
    ... // some magic code mostly the same like in `_captureBackground` method

    return { /* ButtonSprite constructor parameters */ };
}
```

On top of this `ButtonGroup` shouldn't know about shadow thing. We can pass ButtonSprite objects to it.

What do you think @a1exsh ?

a1exsh replied:
Sure, we can do it with a function instead.

The trouble with `ButtonGroup` and adding buttons to it explicitly is that the code to calculate buttons positions is not reused.  I would like us to use the shared code more.

a1exsh replied:
@ihhub well, that's nearly not as easy as it sounded.  Instances of `ButtonBase` or derived classes cannot be copied, due to embedded `unique_ptr`.

With the following code
```cpp
    fheroes2::ButtonSprite btnCancel = fheroes2::makeButtonWithShadow( pos_rt.x + buttonOffset + 129, pos_rt.y + 410, fheroes2::AGG::GetICN( buttonCancelICN, 0 ),
                                                                       fheroes2::AGG::GetICN( buttonCancelICN, 1 ), display );
```
I'm getting this error:
```
../fheroes2/battle/battle_dialogs.cpp: In member function ‘bool Battle::Arena::DialogBattleSummary(const Battle::Result&, const std::vector<Artifact>&, bool) const’:
../fheroes2/battle/battle_dialogs.cpp:501:125: error: use of deleted function ‘fheroes2::ButtonSprite::ButtonSprite(const fheroes2::ButtonSprite&)’
  501 |                                                                        fheroes2::AGG::GetICN( buttonCancelICN, 1 ), display );
      |                                                                                                                             ^
In file included from ../fheroes2/battle/battle_interface.h:33,
                 from ../fheroes2/battle/battle_dialogs.cpp:31:
../fheroes2/gui/ui_button.h:99:11: note: ‘fheroes2::ButtonSprite::ButtonSprite(const fheroes2::ButtonSprite&)’ is implicitly deleted because the default definition would be ill-formed:
   99 |     class ButtonSprite : public ButtonBase
      |           ^~~~~~~~~~~~
../fheroes2/gui/ui_button.h:99:11: error: use of deleted function ‘fheroes2::ButtonBase::ButtonBase(const fheroes2::ButtonBase&)’
../fheroes2/gui/ui_button.h:30:11: note: ‘fheroes2::ButtonBase::ButtonBase(const fheroes2::ButtonBase&)’ is implicitly deleted because the default definition would be ill-formed:
   30 |     class ButtonBase : public ActionObject
      |           ^~~~~~~~~~
../fheroes2/gui/ui_button.h:30:11: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = fheroes2::Sprite; _Dp = std::default_delete<fheroes2::Sprite>]’
In file included from /usr/include/c++/10/memory:83,
                 from ../fheroes2/system/players.h:26,
                 from ../fheroes2/army/army.h:30,
                 from ../fheroes2/battle/battle_dialogs.cpp:27:
/usr/include/c++/10/bits/unique_ptr.h:468:7: note: declared here
  468 |       unique_ptr(const unique_ptr&) = delete;
      |       ^~~~~~~~~~
```

Any idea how to overcome it?

ihhub replied:
Ah, we have to define move-constructor and move operator but delete copy-constructor and copy operator for ButtonSprite class.

a1exsh replied:
@ihhub I managed to convince the compiler, but not sure I can convince myself the code is now better :joy: 

This is getting really out of hand, and is really outside of my C++ capabilities...

ihhub replied:
We all learn something new every day :) I'll take a look.

ihhub reviewed (changes requested):
Hi @a1exsh , I left few more comments in this pull request. Could you please take a look at them and check if they make sense, especially from architectural point of view?

ihhub commented on src/fheroes2/heroes/heroes_spell.cpp:
```diff
@@ -531,16 +531,29 @@ bool ActionSpellTownPortal( Heroes & hero )
+    const fheroes2::Sprite & buttonOkIcn = fheroes2::AGG::GetICN( icnId, 1 );
+    const fheroes2::Sprite & buttonCancelIcn = fheroes2::AGG::GetICN( icnId, 3 );
+
+    std::unique_ptr<fheroes2::ButtonBase> buttonOk(
+        new fheroes2::ButtonSprite( fheroes2::makeButtonWithShadow( area.x + border, area.y + area.height - border - buttonOkIcn.height(), buttonOkIcn,
+                                                                    fheroes2::AGG::GetICN( icnId, 2 ), display ) ) );
+
+    std::unique_ptr<fheroes2::ButtonBase> buttonCancel( new fheroes2::ButtonSprite(
+        fheroes2::makeButtonWithShadow( area.x + area.width - border - buttonCancelIcn.width(), area.y + area.height - border - buttonCancelIcn.height(), buttonCancelIcn,
+                                        fheroes2::AGG::GetICN( icnId, 4 ), display ) ) );
+
+    fheroes2::ButtonGroup btnGroup;
+    btnGroup.createButton( buttonOk, Dialog::OK );
+    btnGroup.createButton( buttonCancel, Dialog::CANCEL );
```
`createButton` method should have this signature
```cpp
void addButton( ButtonSprite && button, const int returnValue )
{
    addButton( new ButtonSprite( std::move( button ) ),  returnValue );
}
```
In this case we don't need unique_ptr usage but can directly invoke `makeButtonWithShadow` call within `addButton` method.

a1exsh replied:
I had the same concerns, but didn't know how to solve it — thanks! :)

ihhub reviewed (changes requested):
Hi @a1exsh , I put few more comments here. I feel we're almost there. Can you please check the comments?

a1exsh commented on src/engine/image.cpp:
```diff
@@ -1803,6 +1847,15 @@ namespace fheroes2
         }
     }
 
+    void ReplaceTransform( const Image & in, Image & out )
```
:heavy_check_mark:

a1exsh commented on src/fheroes2/gui/ui_button.cpp:
```diff
@@ -429,4 +454,47 @@ namespace fheroes2
         }
     }
+
+    ButtonSprite makeButtonWithBackground( int32_t offsetX, int32_t offsetY, const Sprite & released, const Sprite & pressed /*, const Sprite & disabled*/,
```
Oops, I thought I got rid of that one.

a1exsh commented on src/fheroes2/gui/ui_button.h:
```diff
@@ -101,8 +100,13 @@ namespace fheroes2
         ButtonSprite( int32_t offsetX = 0, int32_t offsetY = 0 );
         ButtonSprite( int32_t offsetX, int32_t offsetY, const Sprite & released, const Sprite & pressed, const Sprite & disabled = Sprite() );
+        ButtonSprite( ButtonSprite & ) = delete;
+        ButtonSprite( ButtonSprite && o ) noexcept;
```
OK

a1exsh commented:
@ihhub it complies and works, but I will not pretend I understood this `std::move()` magic :sweat_smile:

ihhub commented on src/fheroes2/gui/ui_button.cpp:
```diff
@@ -429,4 +479,45 @@ namespace fheroes2
+        Sprite disabled( released );
+        ApplyPalette( disabled, PAL::GetPalette( PAL::PaletteType::DARKENING ) );
+
+        Sprite disabledWithBackground( croppedBackground, 0, 0 );
```
```cpp
Sprite disabledWithBackground( std::move( croppedBackground ) );
disabledWithBackground.setPosition( 0, 0 );
```

a1exsh replied:
This does feel like a micro-optimization, but doesn't seem to hurt readability too much. ;)

ihhub replied:
Haha. It's actually a huge performance boost as we don't make system calls to allocate and free memory on stack.

ihhub commented on src/fheroes2/heroes/heroes_spell.cpp:
```diff
@@ -531,16 +531,24 @@ bool ActionSpellTownPortal( Heroes & hero )
-    btnGroups.draw();
+    const int32_t border = 10;
+    const int icnId = isEvilInterface ? ICN::SYSTEME : ICN::SYSTEM;
+    const fheroes2::Sprite & buttonOkIcn = fheroes2::AGG::GetICN( icnId, 1 );
+    const fheroes2::Sprite & buttonCancelIcn = fheroes2::AGG::GetICN( icnId, 3 );
```
These aren't ICNs but images. It's good to call them `buttonOkImage` and `buttonCancelmage`.

ihhub commented on src/fheroes2/heroes/heroes_spell.cpp:
```diff
@@ -512,14 +512,14 @@ bool ActionSpellTownPortal( Heroes & hero )
         return false;
     }
 
-    std::unique_ptr<fheroes2::StandardWindow> frameborder( new fheroes2::StandardWindow( 290, 250 ) );
```
I remember that was done on purpose to avoid some drawing issues. Can you please explain why are we changing it?

a1exsh replied:
@ihhub you are right, we should not change it.  I now see why it was done, after consulting the commit history: https://github.com/ihhub/fheroes2/commit/438257c7b7f7b206c319c514d83c2d83dc3f801c

I will put it back and add a comment for the next developer who comes across it and thinks we don't need the pointer there :D

ihhub replied:
Great!

ihhub commented on src/fheroes2/dialog/dialog_giftresources.cpp:
```diff
@@ -230,9 +230,21 @@ void Dialog::MakeGiftResource( Kingdom & kingdom )
-    btnGroups.draw();
+    const int32_t border = 10;
+    const int icnId = Settings::Get().ExtGameEvilInterface() ? ICN::SYSTEME : ICN::SYSTEM;
+    const fheroes2::Sprite & buttonOkIcn = fheroes2::AGG::GetICN( icnId, 1 );
+    const fheroes2::Sprite & buttonCancelIcn = fheroes2::AGG::GetICN( icnId, 3 );
```
These aren't ICNs but images. Shall we call them `buttonOkImage` and `buttonCancelImage` respectively?

a1exsh replied:
True.  They are actually Sprites, so called them like that.

ihhub reviewed (changes requested):
Hi @a1exsh , I left just few more optimization related comments. Could you please check them too?

ihhub commented:
> @ihhub it complies and works, but I will not pretend I understood this `std::move()` magic 😅

Move semantics are used to avoid memory copy and only for objects which are going to be destroyed just after the call. This is useful when you can just swap values instead of copy them. In case of images it's essential to avoid extensive memory manipulations.

a1exsh commented:
> Move semantics are used to avoid memory copy and only for objects which are going to be destroyed just after the call.

Well, that is clear.  What I don't get is why in some places it's OK to just call `new ButtonSprite( ... )` and in other places you have to specify `std::move()` explicitly...

ihhub commented:
> Well, that is clear. What I don't get is why in some places it's OK to just call `new ButtonSprite( ... )` and in other places you have to specify `std::move()` explicitly...

std::move enforces in some sense to end object's lifetime within your code block. If you don't do this then you're passing by normal reference rather than by r-value.

ihhub commented:
@Branikolog please retest the latest build again as we changed a lot of internal code.

a1exsh commented:
appveyor build failed with an internal error, please re-trigger:
```
MSBUILD : error MSB4017: The build stopped unexpectedly because of an unexpected logger failure.
Microsoft.Build.Exceptions.InternalLoggerException: The build stopped unexpectedly because of an unexpected logger failure. ---> Microsoft.Build.Exceptions.InternalLoggerException: The build stopped unexpectedly because of an unexpected logger failure. ---> System.Net.WebException: The remote server returned an error: (500) Internal Server Error.
```

ihhub commented:
@a1exsh , a huge thanks for these changes!
